### PR TITLE
docs: Klip-5: Internal Schemas for KSQL Engine

### PR DESCRIPTION
**Pull request by hjafarpour**
# KLIP-5: Internal Schemas for KSQL Engine

 **Author**: hjafarpour |

 **Release Target**: 5.4 |

 **Status**: In Discussion |

 **Discussion**:

 **tl;dr:** *Introduce the internal schemas for KSQL engine to decouple the internal identifier names from the
external field names so KSQL can handle any field name in the input data.*

 ## Motivation and background

 Currently when users declare the schema for a stream or table, KSQL engine uses the provided schema
to perform query processing on the stream or table data. This means that the engine uses the same
names in the schema fields to refer to the schema fields in computations such as expression evaluation
 or filter evaluation. We generate Java code to evaluate such expressions which imposes certain
 restrictions on variable naming. However, requiring user schemas to comply with variable naming
 requirements in Java limits the KSQL ability to process any user data. Although in Avro and Delimited
 formats KSQL won’t have any issues since the schema field names already are compatible with the current
 KSQL naming requirements, in handling content with JSON format, KSQL would encounter limitations.
 In JSON format, field names don’t necessarily conform to KSQL naming requirements and a valid JSON
 message can have field names that cannot be accepted by KSQL. In addition to the above requirement,
 the field names in KSQL should also comply with the identifier naming requirements in SQL in order to
 have the statements parse correctly.

 In this KLIP, we propose to decouple the internal schema, that KSQL uses for query processing, from
 the external schema that user content conforms to. The internal KSQL schema will be fully compliant
 with KSQL requirements and the engine can use it to process queries with no issues. On the other
 hand, external schema won’t be restricted with KSQL requirements and users can use any content with
 any naming convention in KSQL.

 Users will be able to use quoted identifiers to declare and access fields that do not conform to
the KSQL identifier requirements. KSQL will generate an internal schema with a one to one mapping
from the external schema fields into the internal schema fields. All of the computation will use the
field names from the internal schema.

 As an example, consider we want to declare s stream on a topic with values in JSON format. The following
is a sample message value in this topic:

 ```json
{"@ID": 0, "@NAME": "foo", "@AMOUNT": 0.0}
```

 As you can see, two of the field names start with “@” and the third one includes “.” in the field name.
 Currently we cannot declare a stream or table on this topic since its content does not conform to KQSL
 field naming requirements. However, we should be able to use quoted identifiers, which is supported by
 our parser, to declare a stream or table on this topic and write queries over the declared stream or
 table. Let’s assume we declare a stream on this topic, named foo. The following would be the DDL
  statement:

 ```sql
CREATE STREAM foo
  (“@ID” BIGINT, “@NAME” STRING, “MESSAGE.AMOUNT” DOUBLE)
WITH
  (KAFKA_TOPIC=’foo’, VALUE_FORMAT = ‘JSON’);
```

 KSQL will add the above stream with the declared schema to its metastore. Now, let’s assume we want
to create a new stream from foo using the following CSAS statement:

 ```sql
CREATE STREAM bar AS
SELECT “@ID”, lowercase(“@NAME”) AS “@NAME”, “MESSAGE.AMOUNT” * 100 AS “AMOUNT.BY.100”
FROM foo
WHERE “MESSAGE.AMOUNT” < 1;
```

 KSQL will generate an internal schema for foo before processing the above query and rewrites the query
based on the internal schema. It then will use the rewritten query to process the messages from the topic.

 ## Scope

 * Support for Quoted Identifiers in DDL and DML statements
* Creation of internal schema for external schemas
* Query rewrite to use the internal schemas
* Query processing using the internal schemas

 ## Value/Return

 This will eliminate the current requirement of having message field names in Kafka topics to conform
to KSQL internal requirements. With this KLIP, KSQL will be able to process topics with any naming convention.

 ## Public API

 The only user facing part in this KLIP is the support for quoted identifier which will be the same
as the standard quoted identifiers in SQL. The declaration and use of the internal schemas are not
visible to users and are part of the KSQL engine internals.

 ## Design

 The internal schema will only be used in executing DML statements. When the engine receives a DML
statement (CSAS, CTAS, INSERT INTO, SELECT), it builds an internal schema with valid field names for
the given source schema(s). This will be done by extending the current KSQLSchema class by adding
the internal schema to it. The internal schema field names can follow a predefined protocol so that
KSQL engine would build the same internal schema for a given external schema everytime. As an example,
 let’s consider we use the index of the field in the schema as a prefix to a constant string value to
 build the internal schema field names. Consider the previous example where our external schema has
 three fields as the following:

 ```sql
(“@ID” BIGINT, “@NAME” STRING, “MESSAGE.AMOUNT” DOUBLE)
```

 Assuming the constant string prefix to build the internal schema fields is “COL_”, the internal
schema for the above schema will be as the following:

 ```sql
(COL_0 BIGINT, COL_1 STRING, COL_2 DOUBLE)
```

 Note that the following one to one mapping will exist between the fields from the external schema and the fields from the internal schema:

 * “@ID”                            <==>   COL_0
* “@NAME”                     <==>   COL_1
* “MESSAGE.AMOUNT” <==>   COL_2

 In the case of nested fields, we can apply the same approach to the nested fields inside ARRAY, MAP
and STRUCT types.

 Now that we have the internal schema built, the engine will rewrite the queries by replacing the
identifiers that refer to the fields in the external schema with their corresponding field name in
the internal schema.

 As an example, consider the following query again:

 ```sql
CREATE STREAM bar AS
SELECT “@ID”, lowercase(“@NAME”) AS “@NAME”, “MESSAGE.AMOUNT” * 100 AS “AMOUNT.BY.100”
FROM foo
WHERE “MESSAGE.AMOUNT” < 1;
```

 With the above mapping the engine will write this query as the following:

 ```sql
 CREATE STREAM bar AS
SELECT FOO_COL_1 AS “@ID”, lowercase(FOO_COL_1) AS “@NAME”, FOO_COL_2 * 100 AS “AMOUNT.BY.100”
FROM foo
WHERE FOO_COL_2 < 1;
```

 KSQL engine will be able to run the above query since all identifiers conform to the identifier
naming requirements in KSQL. Note that KSQL already appends the source name to the field name
automatically which will eliminate the ambiguity in case of having multiple sources (such as Join queries)
 with the same field names.

 The engine will use the generated internal schema to build the streams application. The deserializers
 from the source stream or table will use the external schema to fetch the fields and build the
 GenericRow object, but beyond this point the internal schema will be used to process the records
 for the queries.

 Also note that the result schema will be generated from the external schemas which means if we select
any field from the external schema in the SELECT expressions and user does not specify an alias for
the result column, the engine will use the same name from the external schema for the corresponding
field in the result schema.

 Note that all the KSQL error or log messages should still use the external schema names since user
does not know about the internal schemas.

 ## Test Plan

 Tests for supporting quoted identifiers will be added.

 All of the existing tests for different stages of query processing will indeed test the correctness
 of the use of internal schemas too since the internal schemas will be used for such processing.

 Tests that validate the intermediate schemas such as the ones in QTT should be updated since the
intermediate schemas will be based on the internal schemas.

 ## Documentation

 Only documentation for the quoted identifier support will be added since that’s the only user facing feature.

 ## Compatibility Implications

 The internal schemas would break compatibility for the stateful queries since the state store
currently stores values based on the external schema names.

 ## Performance Implications

 No performance implication since we just change the identifier names and the processing logic does not change.

 ## Security Implications

 No security implications.

### CODE REVIEW

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+KSQL will generate an internal schema for foo before processing the above query and rewrites the query
+based on the internal schema. It then will use the rewritten query to process the messages from the topic.
+
+## Scope
+
+* Support for Quoted Identifiers in DDL and DML statements
+* Creation of internal schema for external schemas
+* Query rewrite to use the internal schemas
+* Query processing using the internal schemas
```
Comment (big-andy-coates) [inline design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: * Quoting a field name should also make it case sensitive, right?  Is that in scope?

Unquoted identifiers are case-insensitive, while quoted identifiers are case-sensitive. The SQL spec defines comparison rules for identifiers. If both sides are unquoted then they are always case-insensitive, e.g. column_name == ColumN_NamE. However quoted identifiers are case-sensitive, therefore "table_name" != "TAble_naME". Unquoted and quoted identifiers can only be considered the same if the quoted characters are uppercased, e.g. TABLE_NAME == "TABLE_NAME", but Table_Name != "Table_Name".

Here is the relevant part of the spec (section 5.2.13):

         13)A <regular identifier> and a <delimited identifier> are equiva-
            lent if the <identifier body> of the <regular identifier> (with
            every letter that is a lower-case letter replaced by the equiva-
            lent upper-case letter or letters) and the <delimited identifier
            body> of the <delimited identifier> (with all occurrences of
            <quote> replaced by <quote symbol> and all occurrences of <dou-
            blequote symbol> replaced by <double quote>), considered as
            the repetition of a <character string literal> that specifies a
            <character set specification> of SQL_TEXT and an implementation-
            defined collation that is sensitive to case, compare equally
            according to the comparison rules in Subclause 8.2, "<comparison
            predicate>".

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+As you can see, two of the field names start with “@” and the third one includes “.” in the field name.
+ Currently we cannot declare a stream or table on this topic since its content does not conform to KQSL
+ field naming requirements. However, we should be able to use quoted identifiers, which is supported by
+ our parser, to declare a stream or table on this topic and write queries over the declared stream or
+ table. Let’s assume we declare a stream on this topic, named foo. The following would be the DDL
+  statement:
+
+```sql
+CREATE STREAM foo
```
Comment (big-andy-coates) [inline design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: I think the parser supports back quotes for quoting too.  We should either disable or support & document.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+## Public API
+
+The only user facing part in this KLIP is the support for quoted identifier which will be the same
+as the standard quoted identifiers in SQL. The declaration and use of the internal schemas are not
+visible to users and are part of the KSQL engine internals.
+
+## Design
+
+The internal schema will only be used in executing DML statements. When the engine receives a DML
```
Comment (big-andy-coates) [inline design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: > This will be done by extending the current KSQLSchema class by adding the internal schema to it.

I recommend we do this by refactoring `LogicalSchema` to no longer use Connect's `Schema` type internally. (I've been doing work to move us in this direction anyway).  

Our own Field class could:
- know if the field is quoted or not, without _encoding_ the quotes into a string).
- Handle field name comparison, taking into account case-sensitivity of quoted identifiers and the SQL standard for them. (It's sufficient to ensure unquoted field names are upper cased and then treat field names as case sensitive).
- Handle full-qualified fields without _encoding_ the source into the name.
- Track any user defined alias. Meaning deserializers can be passed it to map from source to aliased field names.
- Track the new 'internal' field name.
- Use the KSQL type system, rather than Connects, i.e. make use of `SqlType` enum and something like `io.confluent.ksql.parser.tree.Type`s subclasses.

This has many benefits over trying to bolt this onto/into the Connect schema.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 1 above)
Comment (big-andy-coates) [inline design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: * Quoting a field allows it to be a reserved work, right? Is that in scope?

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+any field from the external schema in the SELECT expressions and user does not specify an alias for
+the result column, the engine will use the same name from the external schema for the corresponding
+field in the result schema.
+
+Note that all the KSQL error or log messages should still use the external schema names since user
+does not know about the internal schemas.
+
+## Test Plan
+
+Tests for supporting quoted identifiers will be added.
```
Comment (big-andy-coates) [inline design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: With the ability to quote identifiers, it would now be possible for the user to create field names that are not valid for the serialization format being used.  We should ensure this results in an error when the statement is submitted, rather than just result in runtime serde errors.

**Comment by hjafarpour** [comment]
Thanks for the comments @agavra 

- ` How do we handle case sensitivity in the mappings? This isn't impossible, but I think a section on this is important for completeness.`
The quoted identifiers are case sensitive and the unquoted ones are case insensitive. The mapping does not care about case sensitivity, it rather cares about mapping an identifier, quoted or not, to the correct internal field.

- `You mention that KSQL errors should use external schema names. I think this is very important, it would be nice if there was a safeguard to make sure that these names are never leaked to the top level.`
Yes, we should not expose the internal names to user. I don't have any specific safeguard in mind other than making sure the error messages use the external names but we may standardize the error message creation and enforce this there. Let me know if you have any suggestions on this.

- `If an identifier is unquoted, do we treat it differently? What error will we throw if an unquoted identifier is invalid?`
Unquoted identifiers will be treated as we do today. They have to follow up KSQL requirements for identifier naming and if there is any error, the provided external name will be used in the error message.

- `Do we do any validation with serializers/deserializers? This will make KSQL less restrictive than some of the serialization formats. How do errors work there?`

@big-andy-coates  also brought this up. We should enforce the format requirements before starting the query. For instance, if we read from JSON format and write into AVRO format we should fail if the output schema is not valid in Avro. This should be done in the query compile time and not in the runtime.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 2 above)
Comment (hjafarpour) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: Yes, for now I think we can disable back quote and only support double quotes. We can easily add the support for back quotes if we want to later.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 1 above)
Comment (hjafarpour) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: Yes, the quoted identifiers are case sensitive. Note that in this KLIP we only consider the identifiers for field names in schemas and identifiers for stream/table names are out of scope of this KLIP. I updated the KLIP to reflect this.

Comment (hjafarpour) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: Yes, quoting a fields allows you it to have any string value in the quotes. You can have reserved words in the quotes and handling it is in the scope of this KLIP. For instance we should be able to handle the following:
```sql
SELECT "SELECT" from foo;
```

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 3 above)
Comment (hjafarpour) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: Totally agree, you have been making lot's of changes on handling schemas and fields and the implementation of this KLIP will use the latest changes. On having our own `Field` class too I agree it would make the implementation of this KLIP much easier.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+Now that we have the internal schema built, the engine will rewrite the queries by replacing the
+identifiers that refer to the fields in the external schema with their corresponding field name in
+the internal schema.
+
+As an example, consider the following query again:
+
+```sql
+CREATE STREAM bar AS
+SELECT “@ID”, lowercase(“@NAME”) AS “@NAME”, “MESSAGE.AMOUNT” * 100 AS “AMOUNT.BY.100”
```
[the remark this answers is not in the document]
Comment (hjafarpour) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: I updated the example and used fully qualified name for the second expression.
Yes, if there is no alias specified for a select expression with single quoted identifier, the alias will be the same quoted identifier. The example indeed includes this for the first expression:
I updated the KLIP and added the generated sink schema too.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+FROM foo
+WHERE “MESSAGE.AMOUNT” < 1;
+```
+
+With the above mapping the engine will write this query as the following:
+
+```sql
+
+CREATE STREAM bar AS
+SELECT FOO_COL_1 AS “@ID”, lowercase(FOO_COL_1) AS “@NAME”, FOO_COL_2 * 100 AS “AMOUNT.BY.100”
```
[the remark this answers is not in the document]
Comment (hjafarpour) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: Yes, but now the prefixing should be done as part of the rewrite in this section since we will have ambiguity if we have more than one source, for instance in JOIN queries.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 4 above)
Comment (hjafarpour) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: Exactly, the engine should enforce the requirements of the sink format. For instance, if the sink is AVRO and there is a fields name that is not valid for avro the engine should enforce this in compile time, not runtime. I will clarify this in the KLIP.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+This verification should be done in query compile time. For instance, if the sink format is Avro, and the
+field names in the result schema do not comply with Avro naming requirments, KSQL engine should not run
+the query.
+
+Note that all the KSQL error or log messages should still use the external schema names since user
+does not know about the internal schemas.
+
+## Test Plan
+
+Tests for supporting quoted identifiers will be added.
```
Comment (big-andy-coates) [inline design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: Other thoughts for the test plan:

1. Ensure we have test coverage for quoted identifiers that include periods in their name, e.g. `FOO.x`.  We currently _encode_ the source name into fields and this is likely to either not work, or be really ambiguous / error prone if we start to allow periods in identifiers.  Hence my recommendation that we implement our own Field type with explicit handling of fully qualified names.

2. Ensure we have test coverage of queries that select fields named ROWKEY and ROWTIME.  At the moment `KsqlStructuredDataSourceNode` will filter any such fields out.  We'll have to re-think this functionality if we're to support something like:

```sql
CREATE STREAM x AS SELECT "ROWTIME", y AS "ROWKEY" FROM Z;
```

We may choose to initially allow quoted identifiers to be reserved words, except `ROWTIME` and `ROWKEY`. But this will need documenting / testing.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: true dat.

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 3 above)
Comment (big-andy-coates) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: Probably worth mentioning this in the KLIP...

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 1 above)
Comment (big-andy-coates) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: You've updated the KLIP to call out its only looking to support quoted field names. But you've not updated the KLIP about the case-sensitivity, which is what my comment was about...

Comment (big-andy-coates) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: So can we update the KLIP to include this in the scope?

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: > I updated the example and used fully qualified name for the second expression.

I don't see this - the KLIP still shows 

```sql
SELECT FOO."@ID", lowercase("@NAME") etc.
```

Not

```sql
SELECT FOO."@ID", lowercase(FOO."@NAME") etc.
```

???

File: design-proposals/klip-5-internal-schemas-for-KSQL-engine.md (hunk 7 above)
Comment (hjafarpour) [reply design-proposals/klip-5-internal-schemas-for-KSQL-engine.md]: We should be able to use `"ROWKEY"` and `"ROWTIME"` since they are the same as any other reserved word and as long as they are quoted KSQL should be able to use treat them as any other quoted identifier.
